[sully] 페어 리뷰용 PR

devbattery opened the pull request.

Commit:
docs: 학습 계획 작성

Commit:
feat: 마크업 파일 업로드

블로그에 있던 양식을 참고하여 root 폴더에 바로 놨는데 이게 맞는지는 잘 모르겠다

Commit:
fix: 마크업 파일 제거

프로젝트 구조를 공부하고 보니 resources 폴더 내에 static 폴더로 마크업 파일이 존재한다는 것을 깨달았다.

Commit:
feat: 카카오 마크업 파일로 변경

원래 위치한 static 폴더는 미션에서 주어진 디자인이 아니었다.

Commit:
feat: template 파일인 mustache 적용 시도

mustache 플러그인 설치 후 templates 폴더 안에 파일이 위치시키게 했다. 나중을 위한 발판이다.

Commit:
fix: 기존 qna 폴더로 다시 수정

호눅스와의 클래스를 갖고 난 후 제일 쉬운 방법이 기존의 폴더 방식이라는 것을 알게 되어서 기본 코드에 있는 폴더로 다시 수정했다.

Commit:
feat: templates 폴더에 form.html, index.html 복사 후 추가

동적 회원가입 기능 구현을 위해 복사

Commit:
feat: 타임리프 설정 후 폴더 위치 변경

gradle과 각 html 파일에 타임리프 코드를 추가해주고 폴더 위치를 static에서 templates로 옮겼다.

Commit:
feat: 사용자가 전달한 값을 관리하는 User 클래스 생성

각 필드를 생성한 후 setter와 getter 생성

Commit:
fix: userId를 id로 변경

내가 한 것은 index였고, 여기서 해야 하는 것은 아이디 그 고유의 값이므로 String형으로 변경

Commit:
style: id의 변수명 변경

html 파일을 보다 보니 id가 아닌 userId로 작성되어 있는 것을 발견
변수명만 간단하게 변경해 줌

Commit:
feat: 회원가입 버튼을 누르면 form.html로 넘어가게끔 구현

이제 코드 중복을 제거하고 매핑하는 데 문제가 없도록 할 예정입니다.

Commit:
feat: 가장 하단에 위치한 중복 코드 분리

header.html이라는 네이밍

Commit:
feat: 가장 상단에 위치한 중복 코드 분리

header.html이라는 네이밍

Commit:
feat: footer와 header를 불러오도록 구현

th:replace를 이용하여 분리한 코드를 불러옴

Commit:
frag: navbar-collapse 코드 분리 후 적용

index.html에서 user/join으로 수정해줬던 그 코드이다. 모든 코드에서 중복이 되어서 다 수정했다.

Commit:
fix: "."을 제거하였고, footer로 되어 있던 것을 수정

</div>.로 되어 있어서 그 부분을 수정하였고
index.html에서 navbar-collapse.html를 불러와야 할 부분에 footer.html이 불러오도록 짜여져 있었음

Commit:
style: navbar-collapse에서 navbar-collapse1로 변수명 변경

navbar-collapse2를 위해 변경

Commit:
frag: navbar-collapse1 코드 분리 후 적용

navbar-collapse2와 똑같은 방식이다

Commit:
frag: subnav.html 코드 분리 후 적용

Commit:
frag: navbar-fixed-top.html 코드 분리 후 적용

Commit:
feat: 유저 리포지토리 인터페이스 생성

MemoryUserRepository에서 이 인터페이스를 구현

Commit:
feat: save 메서드의 반환값을 Optional<>로 변경

값이 안 들어올 수 있음

Commit:
feat: UserRepository를 구현하는 MemoryUserRepository 클래스 생성 후 save 메서드 구현

Map의 key는 userId, value는 User 객체를 받는 HashMap을 생성
단, 실무에서는 동시성 문제 때문에 ConcurrentHashMap을 사용하지만 db에서 없애줄 예정이므로 HashMap 사용
userMap에 넣어줄 때 이미 key값이 존재한다면 Optional의 empty를 반환하도록 함

Commit:
feat: find 역할의 기능 추가

UserId와 객체를 반환해주는 메서드 두개를 구현했다. 항상 userId는 null인 경우를 생각한다.

Commit:
fix: get 방식에서 post 방식으로 변경

Commit:
feat: UserRepository 객체를 관리하는 Service 생성

UserRepository 객체를 생성자로 받아와서 관리한다.
userId가 중복되는지 판별하고, 저장해주는 회원가입 기능을 갖고 있다.
user 전체를 찾는 기능도 갖고 있다.

Commit:
feat: 회원가입 할 때 입력받은 값들을 받아올 수 있도록 생성자 생성

model에 저장한 값을 User 클래스 안에 넣어줄 예정이다.

Commit:
feat: UserService 객체를 관리하도록 생성자 추가

Commit:
feat: 사용자 추가 기능 구현

전에 User와 UserForm을 수정한 이유가 바로 이것이다. 사용자를 받아서 users에 리다이렉트 되도록 한다.

Commit:
feat: 리다이렉트된 /users를 list.html에 리턴되도록 구현

이게 가능한 이유는 상대 경로로 넘겨줄 때 ViewResolver가 경로를 추가해주기 때문

Commit:
feat: model로 받아온 리스트를 각각 테이블에 전달

하지만 userId, name, email에 노란색 밑줄로 Cannot resolve가 뜨는데 정상적으로 넘어옴. 이 이유는 모르겠음

Commit:
fix: HTML 코드 중복 분리 수정

form에서만 header부분의 css 경로가 살짝 달랐고, navbar-fixed-top.html에서 빼먹었던 부분을 추가했다.

Commit:
feat: profile.html로 userId의 user 객체를 넘겨주는 메서드 추가

디버깅으로 어쩌구/users/입력한ID 입력하니 잘 넘어오는 것을 확인

Commit:
fix: 경로 수정

Commit:
fix: header 분리

header와 header2로 분리하여 각 파일에 맞는 frag로 정의했다

Commit:
fix: header2 경로 수정

예전과 똑같은 실수함. ../css로 변경

Commit:
frag: index와 index가 아닌 html 분리

index에서 접근할 때는 "/user/어쩌구"가 맞지만
거기로 이동한 후에는 "/어쩌구"가 맞다.
이렇게 하면 로그인으로 이동한 후에 계속 로그인을 클릭해도 에러가 뜨지 않는다.

Commit:
fix: redirect 이해 후 적용

redirect를 잘 이해하지 못 하고 있어서 Direct 메서드에서 그냥 /login에 가게 만드는 실수를 저질렀다.
직접 몇번 실험해 보면서 이럴 때 redirect를 쓰는 것임을 깨달았다.

Commit:
fix: 절대 경로로 수정

상대 경로로 했기 때문에 header2, for-index와 같은 파일들이 추가적으로 생성할 수밖에 없었다. 절대 경로로 수정한 이후로는 한 종류의 html 파일로만 적용이 가능하게 됐다

Commit:
fix: Direct 메서드 삭제

html에서 상대 경로로 전달해줬었기 때문에 상대경로/상대경로 이런 식으로 계속 경로 추가가 됐다. 하지만 절대 경로로 바꾸면 이 모든 것이 해결된다.

Commit:
feat: 회원가입 후 프로필로 넘어가는 기능 구현

회원가입 했을 때의 userId를 바탕으로 하여 경로에 추가해준다. 그러면 그 userID의 profile로 연결이 된다.

Commit:
feat: userId와 email 연결

profile.html에 model이 넘겨줬으니 그 정보를 바탕으로 출력

Commit:
fix: header2, for-index 없앤 결과 적용

header2를 없애고, for-index를 없앴으니 그것을 user 폴더 내의 파일과 index에도 적용시켜야 함

Commit:
fix: save 메서드의 반환 타입을 void로 변경

return 값이 존재할 이유가 아직까지는 없다.

Commit:
feat: 사용자 아이디를 클릭하면 그 프로필로 이동하게끔 구현

Posts의 각 게시물마다 작성자를 클릭하면 이처럼 되도록 구현 예정

Commit:
frag: qna 폴더 파일 templates로 이동 후 중복 분리

Commit:
feat: 각 파일에 qna 연결 후 경로 매핑

WebMVCConfigurer를 사용해서 매핑 방법 바꿔보는 것도 생각하기

Commit:
feat: 글쓴이, 제목, 내용 정보를 담고 있는 Article 클래스 생성

Commit:
feat: 글쓴이, 제목, 내용 정보를 처리하는 dto 생성

Commit:
feat: Article 정보를 저장하고, 찾는 인터페이스 생성

Commit:
feat: ArticleRepository를 구현하는 클래스 생성

Commit:
feat: ArticleRepository를 객체로 갖고 관리하는 서비스 생성

Commit:
style: 변수 이름 변경

write와 contents로 통일해주었다. html 파일의 id명을 참고했다.

Commit:
style: Form의 변수 이름 변경

write와 contents로 통일해주었다. html 파일의 id명을 참고했다.

Commit:
feat: 글 쓴 정보를 메인 페이지로 리다이렉트 하도록 구현

MvcConfig를 사용하려 했지만 잘 되지 않아 TODO로 남겨두었다.
글 썼던 정보는 form으로 다 잘 넘어온다.

Commit:
feat: WebMVCConfigurer를 적용하여 추가

WebMvcConfigurerAdapter가 deprecated 되었으므로 WebMvcConfigurer로 사용

Commit:
fix: "/"로 GetMapping 받아서 ArticleController에서 오류 났던 것 수정

Commit:
feat: index.html로 List 보내는 메서드 작성

Commit:
feat: /questions/post를 PostMapping 하기 위해 추가

Commit:
feat: index에 게시물이 계속 나열되도록 기능 추가

클릭하면 articles/아이디의 링크로 이동하므로 이에 대한 수정이 필요함

Commit:
fix: 게시물을 클릭하면 index 경로로 들어가게끔 설정

게시물을 추가할수록 index 1,2,3,4 이런 식으로 1씩 증가한다

Commit:
feat: Long형 변수 articleIndex를 추가

레포지토리, 서비스의 관계에 articleIndex가 엮일 수 있도록 했다.
게시물이 실행되면 articleIndex가 1씩 추가되도록 하였다.

Commit:
feat: show.html에 articleIndex를 넘겨주도록 구현

이를 위해 findByArticleIndex 메서드를 만들었음

Commit:
feat: 게시물에 관련된 정보를 표시할 수 있도록 타임리프 문법 추가

Commit:
feat: subnav에 질문하기 버튼 추가

active 타입이 고정되어 있어서 해결하려 했지만 현재 내 css 실력으로는 불가능

Commit:
feat: 수정 정보를 입력하는 폼 생성

form.html를 복사하여 수정하였다.

Commit:
feat: 수정 버튼을 누르면 userId의 경로로 가게끔 추가

정확히는 users/update/{userId}의 경로로 가게 했다.

Commit:
feat: 수정 버튼을 누르면 그 userId에 맞는 경로로 이동하도록 구현

Commit:
fix: label 추가

Commit:
style: button 네이밍 변경

회원가입에서 수정으로 변경

Commit:
feat: MvcConfig를 컨트롤러에 적용

MvcConfig는 WebMvcConfigurer를 구현한 Configuration 애노테이션을 사용하는 클래스

Commit:
bug: 회원 정보 수정하는 기능 추가 후 버그

수정한다는 개념이 제대로 머리에 박히질 않는다. 이론 공부 후 다시 구현해야겠다.

Commit:
fix: Optional 에서 get() 하는 것은 isPresent() 체크 필요

Commit:
fix: 불필요한 코드 제거

set을 한 번 더 해줄 필요가 없었음

Commit:
feat: put 방식 추가 후 타임리프 링크 설정

Commit:
delete: Article 클래스의 불필요한 세터 삭제

Commit:
feat: JdbcTemplate, H2 DB 추가

Commit:
feat: jdbc와 h2 추가에 따른 properties 수정

Commit:
refactor: 피드백에 따라 articleIndex를 id로 수정

Commit:
refactor: userId 기반을 id 기반으로 변경

Commit:
refactor: userId 기반을 id 기반으로 변경 (UserForm 까먹음)

Commit:
feat: h2 데이터베이스 연결 프로퍼티 작성

Commit:
fix: save 메서드 수정

id 중심으로 수정해 주었으니 save 메서드도 id가 증가하는 형식으로 바꿔줌

Commit:
feat: runtimeOnly를 implementation으로 변경

Commit:
feat: 내장 h2 연결

내가 했던 건 서버 h2 연결이었음

Commit:
feat: 회원 정보를 만드는 sql문 생성

id, userId, password, name, email로 이루어졌고, 실행할 때마다 create 된다.

Commit:
refactor: Article 로직에도 Optional 적용

Commit:
feat: 게시글 정보를 만드는 테이블 생성

Commit:
feat(User): Memory가 아닌 DB로 관리하도록 하는 레포지토리 구현

update 메서드는 회원 수정 기능을 따로 구현하면 추가할 예정

Commit:
feat(Article): Memory가 아닌 DB로 관리하도록 하는 레포지토리 구현

Commit:
delete: Memory 레포지토리는 이제 필요 없으므로 삭제

Commit:
replace: sql문으로 더 직관적이게 재작성

Commit:
fix: put 방식에서 post 방식으로 변경 후 수정 가능

Commit:
feat: 회원 정보를 수정하는 update 메서드 구현

Commit:
feat: 400번대 에러 페이지 생성

404, 4xx

Commit:
feat: 스프링 부트 validation 추가

Commit:
refactor: 각각의 Form 객체를 주입받아서 저장하도록 재구현

빌더 방법을 사용하고 싶었지만 천천히 적용시켜볼 생각입니다.

Commit:
feat: 각각의 Form 파일에 @NotBlank 애노테이션 추가

null 값이거나 ""값이 들어오면 4xx 페이지로 넘어가도록 하였습니다.

Commit:
feat: @Validated 애노테이션 추가 후 매개변수 form으로 수정

매개변수로 User나 Article이 들어온 것을 수정했습니다.
Validated를 위해 각자의 Form으로 받아온 뒤 domain 객체에 값들이 들어가도록 했습니다.

Commit:
feat: 회원 수정을 위한 dto 생성

userId 필드를 뺀 상태로 만들었습니다.
이제야 dto에 대한 감이 잡히는 것 같습니다.

Commit:
fix(UserUpdateForm): 회원 수정을 위한 dto 적용

UserForm으로 했을 때는 userId값이 포함되어 있어서 4xx 오류가 났습니다. 하지만 이제는 정상적으로 처리됩니다.

Commit:
feat: UserUpdateForm에 맞는 생성자 추가

이렇게 생성자를 계속 추가해주는 것이 맞는지는 모르겠습니다.

Commit:
refactor: repository와 service의 겹치는 변수명 수정

서비스와 레포지토리에 있는 메서드의 변수명을 각각 다르게 해주었습니다.

Commit:
fix: 에러페이지 수정

사진 추가 후 독립적인 4xx 페이지로 변경

Commit:
fix: 에러 페이지 링크 수정

YES를 누르면 홈페이지
NO를 누르면 블로그

Commit:
feat: sql문에 시간을 저장하는 TIMESTAMP 추가

Commit:
refactor: JdbcTemplate 사용 클래스를 NamedParameterJdbcTemplate와 SimpleJdbcInsert를 사용하도록 리팩토링

Commit:
refactor: user와 article 패키지 분리

Commit:
feat: 현재 시간을 저장하는 LocalDateTime 필드 추가

Commit:
feat(Article): 현재 시간을 원하는 포맷으로 변환한 dto 생성

Commit:
refactor: 패키지 위치 변경

Commit:
refactor(User): 패키지 위치 변경

Commit:
refactor: 컨트롤러에서 도메인을 받지 않도록 재구현

이제 도메인은 서비스에서만 받도록 하였고
컨트롤러는 dto 정보만 받도록 하였다.

Commit:
feat(.html): 현재 시간 정보를 출력하도록 추가

Commit:
refactor(User): JdbcTemplate 사용 클래스를 NamedParameterJdbcTemplate와 SimpleJdbcInsert를 사용하도록 리팩토링

Commit:
feat: 프로필 조회를 위한 dto 생성

Commit:
feat: 회원 수정을 위한 UpdateForm의 생성자 추가

Commit:
refactor: 컨트롤러가 dto를 의존하도록 서비스와 컨트롤러 리팩토링

각 Form이 반환되도록 하였습니다.

Commit:
refactor: profile model을 넘겼으므로 그에 맞춰서 수정

Commit:
refactor(Article): Form의 생성자 매개변수 변경

빌더 패턴은 사용하지 못했지만 Article 객체를 주입해서 처리

Commit:
refactor(User): Form의 생성자 매개변수 변경

빌더 패턴은 사용하지 못했지만 User 객체를 주입해서 처리

Commit:
refactor(User): 겹치는 부분 메서드로 분리 후 스트림 사용

Commit:
refactor(Article): 스트림 사용하여 변환

Commit:
fix(User): 업데이트 기능 수정 완료

PostMapping할 때 매개변수가 User인 게 맘에 걸린다.

Ojeegu commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @GetMapping("/users/{id}")
+    public String findProfile(@PathVariable("id") Long id, Model model) {
```
👀 `@PathVariable` 어노테이션 쓸 때 매개변수 이름이랑 URL 경로 변수 이름 동일한 경우에는 괄호 안에 명시하지 않아도 자동으로 매핑된다고 해요. 이 경우 동일하니까 생략하시는 게 깔끔할 것 같아요.

Ojeegu commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @GetMapping("/users/{id}")
+    public String findProfile(@PathVariable("id") Long id, Model model) {
```
👀 @PathVariable 어노테이션 쓸 때 매개변수 이름이랑 URL 경로 변수 이름 동일한 경우에는 괄호 안에 명시하지 않아도 자동으로 매핑된다고 해요. 이 경우 동일하니까 생략하시는 게 깔끔할 것 같아요.

devbattery replied:
오 그렇네요..! 바로 지웠습니다 감사합니다

Ojeegu commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,64 @@
+        return "user/updateForm";
+    }
+
+    @PostMapping("/users/update/{id}")
+    // TODO: User를 UserUpdateForm으로 바꾸면 오류 남
+    public String postUpdate(@PathVariable Long id, @Validated User user) {
+        userService.updateUser(id, user);
+        return "redirect:/users";
+    }
```
👀 UserService class updateUser 메서드를 보면 매개변수로 User 객체를 받고 있어요. 이걸 UserUpdateForm 객체로 바꾸고, 메서드 내에서 UserUpdateForm -> User 객체로 변환하는 로직을 추가하면 postUpdate에서 UserUpdateForm으로 넘길 수 있을 거예요.

devbattery replied:
저도 그렇게 생각해서 리뷰 전에 이렇게도 해봤는데
똑같이 오류가 나더라구요 ㅠㅠ..

위에 create() 메서드처럼 똑같은 방법으로 한 건데 왜 이것만 오류나는지 모르겠어요

```java
    @PostMapping("/users/update/{id}")
    // TODO: User를 UserUpdateForm으로 바꾸면 오류 남
    public String postUpdate(@PathVariable Long id, @Validated UserUpdateForm form) {
        User user = new User(form);
        userService.updateUser(id, user);
        return "redirect:/users";
    }
```

Ojeegu commented on src/main/java/kr/codesqaud/cafe/controller/dto/article/ArticleTimeForm.java:
```diff
@@ -0,0 +1,61 @@
+        return id;
+    }
+
+    public void setId(Long id) {
+        this.id = id;
+    }
```
👀 ArticleTimeForm이 사용되는 곳이 Service class의 findArticles랑 findArticleId 메서드인 것 같은데 setter가 필요한 곳은 없어보여요...! Articlel 객체에서 값을 받아와서 전달하는 역할로만 사용되는 거라면 setter들은 삭제해도 되지 않을까요?

devbattery replied:
엇 그러네요 게터 세터를 너무 남발한 것 같아요..!
그럼 세터를 다 뺀 DTO는 변수를 final로 선언하는 게 좋을까요?

Ojeegu commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,55 @@
+        return new UserUpdateForm(user);
+    }
+
+    private User findError(Long id) {
+        return userRepository.findById(id)
+                .orElseThrow(() -> new NoSuchElementException("이 아이디를 찾을 수 없어: " + id));
+    }
+
```
👀 메서드 이름이 불분명한 것 같아요..! error 처리는 부수적인거고 일단 id로 user를 찾는거니까 findById 이런 걸로 하는 게 좋지 않을까요?

Ojeegu commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,55 @@
+    private User findError(Long id) {
+        return userRepository.findById(id)
+                .orElseThrow(() -> new NoSuchElementException("이 아이디를 찾을 수 없어: " + id));
+    }
```
👀 메서드 이름이 불분명한 것 같아요..! error 처리는 부수적인거고 일단 id로 user를 찾는거니까 findById 이런 걸로 하는 게 좋지 않을까요?

Ojeegu commented on src/main/java/kr/codesqaud/cafe/controller/dto/user/UserProfileForm.java:
```diff
@@ -0,0 +1,39 @@
+
+    public void setName(String name) {
+        this.name = name;
+    }
```
👀 UserProfileForm도 마찬가지로 setter가 따로 필요 없는 것 같은데 사용하신 이유가 있을까요?

Ojeegu commented on src/main/java/kr/codesqaud/cafe/repository/article/JdbcTemplateArticleRepository.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    @Override
+    public void save(Article article) {
+        SqlParameterSource param = new BeanPropertySqlParameterSource(article);
+        Number key = simpleJdbcInsert.executeAndReturnKey(param);
+        article.setId(key.longValue());
```
👏 덕분에 새로운 방법 알아갑니다

Ojeegu commented on src/main/java/kr/codesqaud/cafe/repository/article/JdbcTemplateArticleRepository.java:
```diff
@@ -0,0 +1,99 @@
+//        return jdbcTemplate.query("select * from articles", articleRowMapper());
+    }
+
+    private RowMapper<Article> articleRowMapper() {
+        return BeanPropertyRowMapper.newInstance(Article.class);
```
👏 깔끔깔끔

Ojeegu commented on src/main/java/kr/codesqaud/cafe/repository/user/JdbcTemplateUserRepository.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Override
+    public void update(Long id, User updateUser) {
+        String sql = "update users " +
+                "set password=:password, name=:name, email=:email " +
+                "where id=:id";
+
+        SqlParameterSource param = new MapSqlParameterSource()
+                .addValue("password", updateUser.getPassword())
+                .addValue("name", updateUser.getName())
+                .addValue("email", updateUser.getEmail())
+                .addValue("id", id);
+
+        template.update(sql, param);
```
👏 따라쳐야겠어요

Ojeegu commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,21 @@
+CREATE TABLE articles
+(
+    id          bigint auto_increment,
+    writer      TEXT      NOT NULL,
+    title       TEXT      NOT NULL,
```
👀 작성자나 글제목은 보통 길이 제한을 두는 편이니까 TEXT보다는 VARCHAR로 선언하는 게 메모리 낭비가 안 될 것 같아요

Ojeegu reviewed (commented):
👏23.04.06 리뷰 끝!

Commit:
fix: 기본 생성자 추가로 오류 수정 (주석 처리)

기본 생성자를 삭제한 후 update한 db를 받아오지 못했었습니다.
또한 문제를 해결하던 도중 동료들이 주석과 같은 로직을 사용하길래 이 방법들이 어떤 차이가 있는지 궁금했습니다.

Commit:
fix: put을 위한 hidden 코드를 주석 처리

현재는 post로 처리하고 있기 때문에 필요하지 않기 때문에 주석처리를 하였습니다.

Commit:
refactor(UserUpdateForm): 컨트롤러가 domain을 의존하던 것을 dto를 의존하도록 수정

그 과정에서 updateUser의 정보를 User에 덮어씌웠습니다.

Commit:
refactor: 단위 테스트를 위해 반환 타입을 변경

void였던 반환 타입을 테스트를 위해 변경해주었습니다.

Commit:
refactor(ArticleController): 메서드명 수정

매핑에서의 메서드명은 정하기 쉽지 않은 것 같습니다.
